### PR TITLE
Replace the 'new collective' form with a modal

### PR DESCRIPTION
**Pull request by mejo-**
### 📝 Summary

* Allows to create a new collective (optionally using an existing circle)
* Allows to select members while creating a collective.
* Resolves: #464

#### 🖼️ Screenshots

Enter name | Select Circle | Select Members
---|---|---
grafik | grafik | grafik

#### Screencast

[recording.webm](https://user-images.githubusercontent.com/3582805/219051767-00d7ab03-46c3-4867-aab7-18d56ec46c6d.webm)

### 🚧 TODO

- [x] Fix cropped bottom in member suggestions
- [x] Fix member types in `onCreate()` before giving `members` to `ADD_MEMBERS_TO_CIRCLE`
- [ ] Cypress tests

### 🏁 Checklist

- [x] Code is properly formatted (`npm run lint` / `npm run stylelint` / `composer run cs:check`)
- [x] [Sign-off message](https://probot.github.io/apps/dco/) is added to all commits
- [ ] Tests (unit, integration and/or end-to-end) passing and the changes are covered with tests
- [x] Documentation ([README](https://github.com/nextcloud/collectives/blob/main/README.md) or [documentation](https://github.com/nextcloud/collectives/blob/main/docs/)) has been updated or is not required

### CODE REVIEW

File: cypress/e2e/pages.spec.js [hunk 1]
```diff
@@ -286,7 +286,7 @@ describe('Page', function() {
 	})
 
 	describe('Display table of contents', function() {
-		it('Allows to display/close TOC and switch page modes in between', function() {
+		it.only('Allows to display/close TOC and switch page modes in between', function() {
```
Comment (juliushaertl) [inline cypress/e2e/pages.spec.js:289]:
```suggestion
		it('Allows to display/close TOC and switch page modes in between', function() {
```

Comment (mejo-) [reply cypress/e2e/pages.spec.js]: Haha, nice catch. Fixed now :blush:

**Review by jancborchardt — DISMISSED** [review]
Looks very nice, very much like in Talk :) cc @marcoambrosini. 2 pieces of feedback:
- [ ] Also like in Talk, in the first step we could already list some permissions, e.g. the editing and sharing permissions, as well as the default page mode? Then it’s not so empty, and the permissions are set before anything is shared, very secure.
- [ ] The "Collective with that name already exists" could already be shown in the name input step, no? (Like how web services often check in the background if the account name is still available.)

**Review by nimishavijay — CHANGES_REQUESTED** [review]
Super nice! Flow looks really great :) Agreed with @jancborchardt plus some points from my side:

- [ ] The icon for selecting a circle seems unclear. Maybe it could a button with a label "Choose circle" or something like that?
- [ ] "Select members" --> "Add people"? cc @jancborchardt on this one
- [ ] It also seems like it's confusing to create a personal collective. We could possibly add a secondary button on the left of "Select members" which says "Create" or something. also cc @jancborchardt 
- [ ] Hover feedback for the emoji seems to be a few pixels too wide, it should be a circle (equal width and height)
- [ ] Each item in the list of users doesn't need to have a background, background shown only on hover and select would work better :)
- [ ] The checkmark could have more margin all around. Even better: Align all checkmarks to the right, similar to creating a new group conversation in Talk 
image
- [ ] Also a question: what is the symbol in the search field on the right? 
image

**Comment by julien-nc** [comment]
Wow that looks very nice! :star_struck: 

* [ ] The bottom right button looks the same when you can't click it (for example "Select members" is disabled when no collective name was entered yet) and when something is loading (for example "Create" is disabled when attempting to create the collective). There could be a loading icon in the button's icon template while creating.
* [ ] There could be an empty content while choosing the collective name, there seems to be one only while choosing the members. The empty content icon could be the app icon and the title could be something like "Enter the new collective name". The icon alone without title would be fine too.

**Comment by mejo-** [comment]
> * [ ]  Also like in Talk, in the first step we could already list some permissions, e.g. the editing and sharing permissions, as well as the default page mode? Then it’s not so empty, and the permissions are set before anything is shared, very secure.

In the long run I thought that we could provide templates for the new collective here. That's why I kept it empty for now. Sure, we could add settings there, but I'd like to do this in a follow-up if at all, as it requires quite a bit of extra work.

>     * [ ]  The "Collective with that name already exists" could already be shown in the name input step, no? (Like how web services often check in the background if the account name is still available.)

Yeah, that would be a nice user experience. On the other hand, that would require an API call to check if a circle with a specifici name exists (circle names are global on the instance) and that would mean to allow enumeration. At the moment, when creating a collective, the backend tries to create the circle and errors out if it already exists. That's way harder to brute-force than an API that allows to check if a name is taken.

**Comment by mejo-** [comment]
>  Also a question: what is the symbol in the search field on the right?

Hehe, that's my password manager :wink:

**Comment by mejo-** [comment]
@nimishavijay @jancborchardt @julien-nc I tried to address all your feedback, thanks a lot. Some ideas I didn't implement, but I commented on them instead, see above. Screenshots and screencast above are updated as well.

**Comment by susnux** [comment]
Looks really nice!
I will look into this later, but from a quick look on the code: Is there a special reason for not using the `NcSelect` component for the member picker? I think it would simplify the code a lot.

**Comment by mejo-** [comment]
> Looks really nice! I will look into this later, but from a quick look on the code: Is there a special reason for not using the `NcSelect` component for the member picker? I think it would simplify the code a lot.

Thanks for looking into it @susnux! The main reason is that the member picker dialog should look similar to how it looks in Talk when creating a new conversation and in Contacts when picking members for a circle. I agree that we should streamline this into a nextcloud-vue component at some point.

File: src/components/Nav/NewCollectiveModal.vue [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+			return this.nameIsTooShort || this.nameIsTaken
+		},
+
+		nameError() {
+			if (this.nameIsTooShort) {
+				return t('collectives', 'Name too short, requires at least three characters')
+			} else if (this.nameIsTaken) {
+				return t('collectives', 'A collective with this name already exists')
+			}
+			return null
```
Comment (max-nextcloud) [inline src/components/Nav/NewCollectiveModal.vue:191]: If you don't return anything `undefined` will be returned - which should be just as good as null here.

```suggestion
```

**Review by max-nextcloud — APPROVED** [review]
Just had a look at the code so far.
Looks good to me. :tada: 
Just some minor comments.

**Review by nimishavijay — APPROVED** [review]
Looks great! Only point I noticed is the button for choosing a circle is a bit confusing without a label, but we can merge this now and open a follow up issue about it :)

File: src/components/Member/MemberPicker.vue [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+			class="selected-members">
+			<NcUserBubble v-for="member in selectionSet"
+				:key="member.key || `member-${member.type}-${member.id}`"
+				:margin="0"
+				:size="22"
+				:display-name="member.label"
+				class="selected-member-bubble">
+				<template #title>
+					<a href="#"
+						:title="t('collectives', 'Remove {type} {name}', { type: member.type, name: member.label })"
```
[the remark this answers is not in the document]
Comment (mejo-) [reply src/components/Member/MemberPicker.vue:27]: Agreed, good catch :ok_hand:

File: src/components/Member/MemberPicker.vue [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+			} catch (e) {
+				console.error(e)
+				showError(t('collectives', 'An error occured while performing the search'))
+				this.membersLoading = false
+				return
+			}
+
+			const data = response.data.ocs.data
+			const exact = response.data.ocs.data.exact
+			data.exact = [] // removing exact from general results
```
[the remark this answers is not in the document]
Comment (mejo-) [reply src/components/Member/MemberPicker.vue:183]: Uh nice, much better :heart:

File: src/components/Member/MemberPicker.vue [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+			const exactSuggestions = rawExactSuggestions
+				.filter(result => typeof result === 'object')
+				.map(share => this.formatResults(share))
+				// sort by type so we can get user&groups first...
+				.sort((a, b) => a.shareType - b.shareType)
+			const suggestions = rawSuggestions
+				.filter(result => typeof result === 'object')
+				.map(share => this.formatResults(share))
+				// sort by type so we can get user&groups first...
+				.sort((a, b) => a.shareType - b.shareType)
```
[the remark this answers is not in the document]
Comment (mejo-) [reply src/components/Member/MemberPicker.vue:199]: Agreed.

File: src/components/Member/MemberPicker.vue [hunk 6]
```diff
@@ -0,0 +1,338 @@
...
+
+			const allSuggestions = exactSuggestions.concat(suggestions)
+
+			// Count occurances of display names in order to provide a distinguishable description if needed
+			const nameCounts = allSuggestions.reduce((nameCounts, result) => {
+				if (!result.displayName) {
+					return nameCounts
+				}
```
[the remark this answers is not in the document]
Comment (mejo-) [reply src/components/Member/MemberPicker.vue]: This turned out to be rather dead code (leftover from an earlier implementation). The current results don't have a `displayName` at all, their `label` attribute is used. And given that we already group results by user, circle and group, I don't think that we have to take extra measures to distinguish them further. At least not for now.

**Pull request merged**